### PR TITLE
[CN-exec] Fix anonymous records

### PR DESCRIPTION
**Pull request by rbanerjee20**
Previously, anonymous records only properly worked in CN runtime testing when they were the return type of some CN predicate or function. Now anonymous records should work across the board at runtime, getting translated to C structs as before.

~~TODO: Revert CN tutorial examples to use records instead of datatypes and check that it works now.~~
~~TODO: Write small examples of anonymous records that are not the return types of CN predicates/functions and check that they work. Include in CI.~~

### CODE REVIEW

File: backend/cn/lib/cn_internal_to_ail.ml [hunk 1]
```diff
@@ -54,37 +54,15 @@ module MembersKey = struct
...
+    | (id, cn_bt) :: ms, (id', cn_bt') :: ms' ->
       let c = String.compare (Id.s id) (Id.s id') in
-      if c == 0 then
-        0
-      else
+      let c' = BaseTypes.compare (cn_base_type_to_bt cn_bt) (cn_base_type_to_bt cn_bt') in
+      if c == 0 && c' == 0 then
         compare ms ms'
+      else
+        c
```
Comment (dc-mak) [inline backend/cn/lib/cn_internal_to_ail.ml:68]: I think something like this? if the point is lexicographic, then this avoids unnecessary computation
```ocaml
let c = .. in 
if c == 0 then
  let c' = ... in
  if c' == 0 then
    recurse
  else
    c'
else
  c
```

File: backend/cn/lib/cn_internal_to_ail.ml [hunk 2]
```diff
@@ -236,11 +204,12 @@ let rec cn_to_ail_base_type ?(pred_sym = None) cn_typ =
       generate_ail_array bt
       (* TODO: What is the optional second pair element for? Have just put None for now *)
     | CN_tuple ts ->
+      failwith "TODO"
```
Comment (dc-mak) [inline backend/cn/lib/cn_internal_to_ail.ml]: `failwith (__FUNCTION__  ": more detailed explanation")` is usually more helpful to non-devs/users

File: backend/cn/lib/cn_internal_to_ail.ml [hunk 3]
```diff
@@ -2922,6 +2891,7 @@ let cn_to_ail_function_internal
       in
       (bs, Some (List.map mk_stmt ss))
     | Uninterp -> ([], None)
+    (* TODO: Make this throw a clear error. Uninterpreted functions must not exist or be used for runtime checking to work *)
```
Comment (dc-mak) [inline backend/cn/lib/cn_internal_to_ail.ml]: `failwith`? is this likely to be confusing?

File: backend/cn/lib/executable_spec_records.ml [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+  | WrapI (_ity, t) -> add_records_to_map_from_it t
+  | Aligned { t; align } -> List.iter add_records_to_map_from_it [ t; align ]
+  | MapConst (_bt, t) -> add_records_to_map_from_it t
+  | MapSet (t1, t2, t3) -> List.iter add_records_to_map_from_it [ t1; t2; t3 ]
+  | MapGet (t1, t2) -> List.iter add_records_to_map_from_it [ t1; t2 ]
+  | MapDef ((_, _), t) -> add_records_to_map_from_it t
+  | Apply (_pred, ts) -> List.iter add_records_to_map_from_it ts
+  | Let ((_, t1), t2) -> List.iter add_records_to_map_from_it [ t1; t2 ]
+  | Match (e, cases) -> List.iter add_records_to_map_from_it (e :: List.map snd cases)
+  | Constructor (_sym, args) -> List.iter add_records_to_map_from_it (List.map snd args)
```
Comment (dc-mak) [inline backend/cn/lib/executable_spec_records.ml:55]: `IT.fold_` or `IT.fold_subterms`?

File: backend/cn/lib/cn_internal_to_ail.ml [hunk 5]
```diff
@@ -186,26 +159,21 @@ let str_of_bt_bitvector_type sign size =
...
-       sym'
-     | (_, sym') :: _ -> sym')
+let augment_record_map ?cn_sym bt =
+  let sym_prefix = match cn_sym with Some sym' -> sym' | None -> Sym.fresh () in
+  match bt_to_cn_base_type bt with
+  | CN_record members ->
+    (* Augment records map if entry does not exist already *)
+    if not (RecordMap.mem members !records) then (
+      let sym' = generate_sym_with_suffix ~suffix:"_record" sym_prefix in
+      records := RecordMap.add members sym' !records)
```
Comment (dc-mak) [inline backend/cn/lib/cn_internal_to_ail.ml:174]: Might be better to use a Hashtable if you want/need mutability? But do you need mutability?

**Review by dc-mak — COMMENTED** [review]
Few nits/questions, but looks reasonable overall

File: backend/cn/lib/cn_internal_to_ail.ml (hunk 5 above)
Comment (rbanerjee20) [reply backend/cn/lib/cn_internal_to_ail.ml:174]: Shouldn't need mutability - a record type is only ever added to the `RecordMap` in this initial pass through the CN spec and then possibly accessed at some later point during the translation of a record type or at the point of instantiation of an anonymous record.

File: backend/cn/lib/executable_spec_records.ml (hunk 4 above)
Comment (rbanerjee20) [reply backend/cn/lib/executable_spec_records.ml:55]: Thanks, I had a look but I am unsure how to encode what I want to do using (e.g.) `IT.fold_`. Specifically, the `RecordMap` needs to be augmented every time we reach a record, which I have written like this:

```ml
  | ...
  | Record members ->
    (* Anonymous record instantiation -> add to records map *)
    Cn_internal_to_ail.augment_record_map (IT.bt it);
    List.iter (fun (_, it') -> add_records_to_map_from_it it') members
  | ...
```

 but the `Record` case of `IT.fold_` does this:

```ml
  | ...
  | Record members -> fold_list f binders acc (List.map snd members)
  | ...
```

so I can't really intercept this case with the necessary map operation in the same way I think. `IT.fold_subterms` looks similar, just with an optional `bindings` argument and then a call to `fold`.

**Pull request merged**